components/ipfs/ipfs_p3a.cc

```diff
@@ -0,0 +1,114 @@
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "extensions/browser/extension_registry.h"
+
+namespace ipfs {
+
+constexpr size_t kP3ATimerInterval = 5;
+
+// IPFS companion installed?
+// i) No, ii) Yes
+void RecordIPFSCompanionInstalled(content::BrowserContext* context) {
+  const char ipfs_companion_extension_id[] = "nibjojkomfdiaoajekhjakgkdhaomnch";
+  auto* registry = extensions::ExtensionRegistry::Get(context);
+  bool installed =
+      registry->enabled_extensions().Contains(ipfs_companion_extension_id);
+  UMA_HISTOGRAM_BOOLEAN("Brave.IPFS.IPFSCompanionInstalled", installed);
+}
+
+int GetIPFSDetectionPromptBucket(PrefService* prefs) {
+  int bucket = 0;
+  auto infobar_count = prefs->GetInteger(kIPFSInfobarCount);
+  if (infobar_count == 1) {
+    bucket = 1;
+  } else if (infobar_count >= 2 && infobar_count <= 5) {
+    bucket = 2;
+  } else if (infobar_count > 5) {
+    bucket = 3;
+  }
+  return bucket;
+}
+
+// How many lifetime times are IPFS detection prompts shown without installing
+// i) 0 times, ii) 1, iii) 2-5 times, iv) 5+ times or more?
+void RecordIPFSDetectionPromptCount(PrefService* prefs) {
+  const int max_bucket = 4;
+  int bucket = GetIPFSDetectionPromptBucket(prefs);
+  UMA_HISTOGRAM_EXACT_LINEAR("Brave.IPFS.DetectionPromptCount", bucket,
+                             max_bucket);
+}
+
+// IPFS state
+// Ask (0), Gateway (1), Local Node (2), Disabled (3)
+void RecordIPFSGatewaySetting(PrefService* prefs) {
+  auto resolve_method = prefs->GetInteger(kIPFSResolveMethod);
+  UMA_HISTOGRAM_EXACT_LINEAR("Brave.IPFS.GatewaySetting", resolve_method, 4);
+}
+
+// How long did the daemon run?
+void RecordIPFSDaemonRunTime(base::TimeDelta elapsed_time) {
+  if (!elapsed_time.is_zero()) {
+    UMA_HISTOGRAM_CUSTOM_TIMES("Brave.IPFS.DaemonRunTime", elapsed_time,
+                               base::TimeDelta::FromMinutes(1),
+                               base::TimeDelta::FromDays(1), 5);
+  }
+}
+
+IpfsP3A::IpfsP3A(IpfsService* service, content::BrowserContext* context)
+    : service_(service), context_(context) {
+  RecordInitialIPFSP3AState();
+  service->AddObserver(this);
+}
+
+IpfsP3A::~IpfsP3A() {
+  service_->RemoveObserver(this);
+}
+
+void IpfsP3A::RecordInitialIPFSP3AState() {
+  auto* prefs = user_prefs::UserPrefs::Get(context_);
+  RecordIPFSCompanionInstalled(context_);
+  RecordIPFSDetectionPromptCount(prefs);
+  RecordIPFSGatewaySetting(prefs);
+}
+
+void IpfsP3A::RecordDaemonUsage() {
+  FlushTimeDelta();
+  RecordIPFSDaemonRunTime(elapsed_time_);
+}
+
+void IpfsP3A::OnIpfsLaunched(bool result, int64_t pid) {
+  if (timer_.IsRunning()) {
+    timer_.Stop();
+  }
+
+  daemon_start_time_ = base::TimeTicks::Now();
+  timer_.Start(FROM_HERE, base::TimeDelta::FromMinutes(kP3ATimerInterval),
+               base::Bind(&IpfsP3A::RecordDaemonUsage, base::Unretained(this)));
```

Comment:
As discussed in the DM, I think record in `OnIpfsShutdown` is more useful than this periodic 5 min report.

Reply:
Sorry I was wrong here, record =/= report, so having both recording is fine actually.